Move fix_url_for_bad_hosts from Jetpack class to Connection pa…

gititon opened the pull request:
This PR moved fix_url_for_bad_hosts from Jetpack class to Connection package so that we can make the Connection package more independent. It deprecates the old method.

#### Testing instructions:
Ensure all tests continue to pass after this refactoring.

#### Proposed changelog entry for your changes:
Moved Jetpack::fix_url_for_bad_hosts to Connection package

gititon committed:
[not verified] Move fix_url_for_bad_hosts from Jetpack class to Connection package

gititon committed:
Updating invocations of Jetpack::fix_url to use Connection package

lezama commented on class.jetpack-client-server.php:
```diff
@@ -226,7 +227,7 @@ function get_token( $data ) {
 			),
 		);
-		$response = Client::_wp_remote_request( Jetpack::fix_url_for_bad_hosts( Jetpack::connection()->api_url( 'token' ) ), $args );
+		$response = Client::_wp_remote_request( Manager::fix_url_for_bad_hosts( Jetpack::connection()->api_url( 'token' ) ), $args );
```
this looks weird since we are calling the manager, once static and the second time as an instance. should we maike `api_url` call `fix_url_for_bad_hosts` instead?

gititon replied:
The results of api_url() aren’t passed to fix_url… in every context. For that reason, I don’t think we should make api_url() automatically invoke/decorate its return value with fix_url… .
<img width="828" alt="Screen Shot 2019-09-06 at 6 07 08 PM" src="https://user-images.githubusercontent.com/3246867/64463262-5e478b00-d0d1-11e9-8a69-ef4abea3793a.png">

kbrown9 replied:
The `fix_url_for_bad_hosts` method is now in the new Utils class. See a142b9c.

lezama commented on packages/connection/src/Manager.php:
```diff
@@ -1735,4 +1735,24 @@ public function sign_role( $role, $user_id = null ) {
+				return $url;
+			case 'NEVER':
+				return set_url_scheme( $url, 'http' );
+			// default : case 'AUTO' :
```
let's clean this line

lezama commented on packages/connection/src/Manager.php:
```diff
@@ -1735,4 +1735,24 @@ public function sign_role( $role, $user_id = null ) {
+	 * Some hosts disable the OpenSSL extension and so cannot make outgoing HTTPS requsets
+	 */
+	public static function fix_url_for_bad_hosts( $url ) {
+		if ( 0 !== strpos( $url, 'https://' ) ) {
```
let's rewrite this method, I think a simple `Defined JETPACK_CLIENT__HTTPS and JETPACK_CLIENT__HTTPS === JETPACK_CLIENT__HTTPS then set_url_scheme( $url, 'http' )` would achieve the same?

gititon replied:
Did you mean to suggest comparing the constant to itself `JETPACK_CLIENT__HTTPS === JETPACK_CLIENT__HTTPS`, or did you have something else in mind? If you did, I don't follow.

This function doesn't make a ton of sense to me. 
```
public static function fix_url_for_bad_hosts( $url ) {
		if ( 0 !== strpos( $url, 'https://' ) ) {
			return $url;
		}

		switch ( JETPACK_CLIENT__HTTPS ) {
			case 'ALWAYS':
				return $url;
			case 'NEVER':
				return set_url_scheme( $url, 'http' );
		}

		// we now return the unmodified SSL URL by default, as a security precaution
		return $url;
	}
```

It seems to be saying:
1) If we receive a URL without https, return that url.
2) If we receive a URL and `JETPACK_CLIENT__HTTPS` is set to `ALWAYS`, return the url whether or not it's https. I suppose it will necessarily be https at this point since step 1 returned anything non https. At that rate, what is the point of having this case?
3) If we receive a URL and `JETPACK_CLIENT__HTTPS` is `NEVER`, it converts an https url to non https and returns it.
4) Otherwise, it returns the original url. I think this should be the default case in the switch statement instead of a final return statement.

In light of the above, couldn't we rewrite the method as 

```
public static function fix_url_for_bad_hosts( $url ) {
    // If we receive an http url, return it
    if ( 'http' === wp_parse_url( $url, PHP_URL_SCHEME ) ) {
        return $url;
     } 

    // If the url should never be https, ensure it isn't https
    if ( 'NEVER' === JETPACK_CLIENT__HTTPS ) {
        return set_url_scheme( $url, 'http' );
    }

    // Otherwise, return the https url
    return $url
}
```

What do you think?

roccotripaldi replied:
Sounds good, and looks much clearer. Let's use the `Constants` package in the rewrite.

kbrown9 replied:
See a142b9c. The `fix_url_for_bad_hosts` method was rewritten as suggested by @gititon, and the Constants package was used.

gititon committed:
[not verified] Remove unnecessary comment

lezama commented:
> Did you mean to suggest comparing the constant to itself JETPACK_CLIENT__HTTPS === JETPACK_CLIENT__HTTPS

🤣 I meant `JETPACK_CLIENT__HTTPS === 'NEVER'`

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -1735,4 +1735,23 @@ public function sign_role( $role, $user_id = null ) {
+	/**
+	 * Some hosts disable the OpenSSL extension and so cannot make outgoing HTTPS requsets
+	 */
+	public static function fix_url_for_bad_hosts( $url ) {
```
I don't see a good reason for this one to be static. As you can see in the rest of the `Manager`, there aren't many methods that are static. Could we also make this one a non-static one too?

An alternative would be to define this method in another class. Seems like it really doesn't belong anywhere though, which is why I suggested a `Utils` class.

lezama replied:
+1 to `Utils` class

kbrown9 replied:
See a142b9c. The `fix_url_for_bad_hosts` method is now in the new Utils class.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -1735,4 +1735,23 @@ public function sign_role( $role, $user_id = null ) {
+			return $url;
+		}
+
+		switch ( JETPACK_CLIENT__HTTPS ) {
```
When working with constants from a package, it's recommended that we use the `Constants` package for that. Not only this makes our code consistent and independent, but it also protects against fatals.

kbrown9 replied:
See a142b9c. The Constants package was used for the `JETPACK_CLIENT__HTTPS` constant.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -1735,4 +1735,23 @@ public function sign_role( $role, $user_id = null ) {
+	/**
+	 * Some hosts disable the OpenSSL extension and so cannot make outgoing HTTPS requsets
+	 */
+	public static function fix_url_for_bad_hosts( $url ) {
```
Nit: Could we add some php docs here? We try to keep the packaged code up to our highest standards.

kbrown9 replied:
See a142b9c. I added a docblock with param and return.

tyxla commented on packages/connection/src/Client.php:
```diff
@@ -131,7 +131,7 @@ public static function remote_request( $args, $body = null ) {
 
 		$url = add_query_arg( urlencode_deep( $url_args ), $args['url'] );
-		$url = \Jetpack::fix_url_for_bad_hosts( $url );
+		$url = Manager::fix_url_for_bad_hosts( $url );
```
The fact that we're accessing the method from a static method somewhere else is another indication why it shouldn't live in the `Manager`.

tyxla commented on class.jetpack.php:
```diff
@@ -4790,23 +4790,13 @@ public static function api_url( $relative_url ) {
-
-		// we now return the unmodified SSL URL by default, as a security precaution
-		return $url;
+		_deprecated_function( __METHOD__, 'jetpack-7.8', 'Automattic\Jetpack\Connection\Manager' );
```
Let's make sure to include the method too in the third argument, see other deprecated method examples in this same file.

Also, let's be careful about backslashes in single quotes; in order to represent a real backslash, you need to specify them twice in that context:

```suggestion
		_deprecated_function( __METHOD__, 'jetpack-7.8', 'Automattic\\Jetpack\\Connection\\Manager::fix_url_for_bad_hosts' );
```

kbrown9 replied:
See 57bcd35.

tyxla reviewed (commented):
Thanks for working on this 👍 I've left some feedback, let me know what you think.

kbrown9 committed:
Connection: Add Utils class and move fix_url_for_bad_hosts

The connection package has static utility methods that need a place to
live, so let's add a Utils class. Also move the fix_url_for_bad_hosts
method into the new Utils class and update all uses of that method.

kbrown9 commented:
We’re working on moving user authorization into the Connection package, and we’ll need this method. The work that’s been done so far looks great! 

I’ll submit a commit with the changes requested in the reviews.

kbrown9 committed:
Connection: Fix the @package name in the Utils file description.

kbrown9 committed:
Connection: Add backslashes to _deprecated_function argument
